pkg/mcs/tso/server/apis/v1/api.go

```diff
@@ -0,0 +1,153 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apis
+
+import (
+	"net/http"
+	"strconv"
+
+	"github.com/gin-contrib/cors"
+	"github.com/gin-contrib/gzip"
+	"github.com/gin-gonic/gin"
+	tsoserver "github.com/tikv/pd/pkg/mcs/tso/server"
+	"github.com/tikv/pd/pkg/utils/apiutil"
+	"github.com/tikv/pd/server"
+	"github.com/unrolled/render"
+)
+
+// APIPathPrefix is the prefix of the API path.
+const APIPathPrefix = "/"
+
+var (
+	apiServiceGroup = apiutil.APIServiceGroup{
+		Name:       "tso",
+		Version:    "v1",
+		IsCore:     false,
+		PathPrefix: APIPathPrefix,
+	}
+)
+
+func init() {
+	tsoserver.SetUpRestHandler = func(srv *tsoserver.Service) (http.Handler, apiutil.APIServiceGroup) {
+		s := NewService(srv)
+		return s.handler(), apiServiceGroup
+	}
+}
+
+// Service is the resource group service.
+type Service struct {
+	apiHandlerEngine *gin.Engine
+	baseEndpoint     *gin.RouterGroup
+
+	srv *tsoserver.Service
+	rd  *render.Render
+}
+
+func createIndentRender() *render.Render {
+	return render.New(render.Options{
+		IndentJSON: true,
+	})
+}
+
+// NewService returns a new Service.
+func NewService(srv *tsoserver.Service) *Service {
+	apiHandlerEngine := gin.New()
+	apiHandlerEngine.Use(gin.Recovery())
+	apiHandlerEngine.Use(cors.Default())
+	apiHandlerEngine.Use(gzip.Gzip(gzip.DefaultCompression))
+	endpoint := apiHandlerEngine.Group(APIPathPrefix)
+	s := &Service{
+		srv:              srv,
+		apiHandlerEngine: apiHandlerEngine,
+		baseEndpoint:     endpoint,
+		rd:               createIndentRender(),
+	}
+	s.RegisterRouter()
+	return s
+}
+
+// RegisterRouter registers the router of the service.
+func (s *Service) RegisterRouter() {
+	configEndpoint := s.baseEndpoint.Group("/")
+	configEndpoint.POST("/admin/reset-ts", gin.WrapF(s.ResetTS))
+}
+
+func (s *Service) handler() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		s.apiHandlerEngine.ServeHTTP(w, r)
+	})
+}
+
+// ResetTS
+// FIXME: details of input json body params
+// @Tags     admin
+// @Summary  Reset the ts.
+// @Accept   json
+// @Param    body  body  object  true  "json params"
+// @Produce  json
+// @Success  200  {string}  string  "Reset ts successfully."
+// @Failure  400  {string}  string  "The input is invalid."
+// @Failure  403  {string}  string  "Reset ts is forbidden."
+// @Failure  500  {string}  string  "PD server failed to proceed the request."
+// @Router   /admin/reset-ts [post]
+// if force-use-larger=true:
+//
+//	reset ts to max(current ts, input ts).
+//
+// else:
+//
+//	reset ts to input ts if it > current ts and < upper bound, error if not in that range
+//
+// during EBS based restore, we call this to make sure ts of pd >= resolved_ts in backup.
+func (s *Service) ResetTS(w http.ResponseWriter, r *http.Request) {
```

Comment:
It seems that there is no need to duplicate ResetTS() in server/api/admin.go. Some common admin RESTful APIs can be put into pkg/tso/api/tsoadmin.go.

For tso service, we execute the following code somewhere (mightn't be in router.go)
tsoAdminHandler := newTSOAdminHandler(svr, rd)
registerFunc(apiRouter, "/admin/reset-ts", tsoAdminHandler.ResetTS, setMethods(http.MethodPost), setAuditBackend(localLog, prometheus))

Reply:
If it needs more time, let's ignore it for now and check-in this pr first.

Reply:
Very good idea 👍

Reply:
Please review the latest implementation.